### PR TITLE
First pass at get_decimated_surfaces

### LINKED ISSUE #9286: ENH: Add GIfTI output for STC
Wanted to get thoughts from others who may find this feature useful, and insights into possible straightforward ways to implement.  

#### Describe the new feature or enhancement

In an effort to compare/overlay results in fMRI and EEG/MEG, several of us are seeking an easy way to export MNE source results to GIFTI (and potentially NIFTI in the future).  GIFTI would allow other programs (e.g. AFNI, Freesurfer) to then work with source-space data in useful ways (e.g. constrain solutions based on fMRI, make pretty pictures).  

#### Describe your proposed implementation

Add options to mne.SourceSpaces.save( ) or better - a new method .export( ) to allow the export of GIFTI images.  For added convenience, with option for hemisphere, time-window of interest, and overwrite.  Wrapper for nibabel's GIFTI implementation with nuances for writing the relevant data in. 

https://nipy.org/nibabel/reference/nibabel.gifti.html

#### Additional comments
This would be decently useful for folks trying to do EEG-fMRI simultaneous, MEG & fMRI joint analysis, or anyone interested in taking the source estimates from MNE into other softwares (FieldTrip and BrainStorm both have GIFTI imports).

### PR DESCRIPTION
**Pull request by jbteves**
Modifed slightly from this gist:
https://gist.github.com/jbteves/384bfdd3777190f59620c3d988a314fd

Thanks for contributing a pull request! Please make sure you have read the
[contribution guidelines](https://mne.tools/dev/install/contributing.html)
before submitting.

Please be aware that we are a loose team of volunteers so patience is
necessary. Assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.

Again, thanks for contributing!

#### Reference issue
Example: partially closes #9286 

#### What does this implement/fix?
Adds a function that gets the decimated subsurface from a SourceSpaces object.

#### Additional information
Draft: left some notes in-code for parts I'm unsure of. Appears to work with a local test, but not sure how to integrate an automated test for the maintainer's preference (reference data set? Generate synthetic surfaces and ensure that the result is correct?)

### CODE REVIEW

File: mne/source_space.py [hunk 1]
```diff
@@ -2821,3 +2821,57 @@ def compute_distance_to_sensors(src, info, picks=None, trans=None,
...
+    """Get the decimated surfaces from a source space.
+
+    Parameters
+    ----------
+    src: instance of SourceSpaces
+        The object which has decimated surfaces
+
+    Returns
+    -------
+    subsurfaces: list of dicts which contain 'rr' and 'tris' keys
```
Comment (agramfort) [inline mne/source_space.py:2836]:
```suggestion
    surfaces: list of dict
        The decimated surfaces present in the source space. Each dict
        which contains 'rr' and 'tris' keys for vertices positions and
        triangle indices.
```

File: mne/source_space.py [hunk 2]
```diff
@@ -2821,3 +2821,57 @@ def compute_distance_to_sensors(src, info, picks=None, trans=None,
...
+    """
+    # NOTE: is there a better way to do this? Maybe raise a TypeError?
+    assert isinstance(src, SourceSpaces)
```
Comment (agramfort) [inline mne/source_space.py:2839]: see what we do elsewhere. Return a ValueError with a message saying
what is the type expected and what was the type provided.

File: mne/source_space.py [hunk 3]
```diff
@@ -2821,3 +2821,57 @@ def compute_distance_to_sensors(src, info, picks=None, trans=None,
...
+    for surf in src:
+        subsurfaces.append(_get_subsurf(surf))
+
+    return subsurfaces
+
+
+def _get_subsurf(source):
+    """Get the subsurface from a set of vertices; use only one surface at a time!
+    """
+    # NOTE: minimal documentation since private, should I elaborate?
```
Comment (agramfort) [inline mne/source_space.py:2851]: no need

File: mne/source_space.py [hunk 4]
```diff
@@ -2821,3 +2821,57 @@ def compute_distance_to_sensors(src, info, picks=None, trans=None,
...
+    # NOTE: this may be better as a function merged with above
+    # NOTE: is there a better way to say "index"
+    redone = np.ones((sz_orig, 1)) * -1
+    for i in range(len(vertno)):
+        redone[vertno[i]] = i
+
+    new_tris = np.zeros(sub_tris.shape, dtype=np.int32)
+    for i in range(new_tris.shape[0]):
+        for jj in range(new_tris.shape[1]):
+            new_tris[i][jj] = redone[sub_tris[i][jj]]
```
Comment (agramfort) [inline mne/source_space.py:2875]: a similar code is used at

https://github.com/mne-tools/mne-python/blob/main/mne/viz/_3d.py#L1626

and it's also avoiding the for loop. With this new public function we should refactor the 3D viz code using matplotlib backend to use this function

Comment (jbteves) [reply mne/source_space.py:2880]: I'm confused, the code linked does not appear to reindex the triangles such that there is now an independent "tris" field for sub-surfaces like in my (admittedly loop-heavy) implementation.

Comment (jbteves) [reply mne/source_space.py:2881]: Sure, thanks!

**Comment by larsoner** [comment]
The sphinx windows failure is due to https://github.com/sphinx-gallery/sphinx-gallery/pull/904 . I think we can ignore it for now...

**Comment by jbteves** [comment]
Thanks @larsoner ! This makes sense and looks much more efficient, I just did a quick solution for a proof of concept so that's much better.

File: doc/changes/latest.inc [hunk 5]
```diff
@@ -126,6 +126,8 @@ Enhancements
 
 - Add support for reading optical density fNIRS data to :func:`mne.io.read_raw_snirf` (:gh:`10408` by `Robert Luke`_)   
 
+- Added :func:`mne.source_spaces.get_decimated_surfaces` to extract subsurfaces from a SourceSpaces instance or a .fif file (:gh: `10421` by `Joshua Teves`_)
```
Comment (larsoner) [inline doc/changes/latest.inc:129]:
```suggestion
- Added :func:`mne.source_space.get_decimated_surfaces` to extract subsurfaces from a SourceSpaces instance or a .fif file (:gh: `10421` by `Joshua Teves`_)
```

File: doc/changes/latest.inc [hunk 6]
```diff
@@ -126,6 +126,8 @@ Enhancements
 
 - Add support for reading optical density fNIRS data to :func:`mne.io.read_raw_snirf` (:gh:`10408` by `Robert Luke`_)   
 
+- Added :func:`mne.source_space.get_decimated_surfaces` to extract subsurfaces from a SourceSpaces instance or a .fif file (:gh: `10421` by `Joshua Teves`_)
```
Comment (larsoner) [inline doc/changes/latest.inc:129]:
```suggestion
- Added :func:`mne.source_space.get_decimated_surfaces` to extract subsurfaces from a SourceSpaces instance or a .fif file (:gh:`10421` by `Joshua Teves`_)
```

**Review by agramfort — APPROVED** [review]
thx @jbteves @larsoner !

**Comment by agramfort** [comment]
@larsoner merge if happy

**Comment by larsoner** [comment]
Thanks @jbteves !

**Pull request merged**